### PR TITLE
UCT/IB: Remove sighandlers in IB cleanup thread

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**
## What 

Remove sighandlers in cleanup thread.

## Why?

We implement non-blocking IB close using special cleanup thread lacking libc support.
We creating this thread during IB initialization i.e. from ucp_worker_create().
IODEMO installing it's Ctrl-C handler during Demo constructor i.e. before creating cleanup thread which makes this thread inherit the handler.
Handler does logging which crashing in libc-less thread.

### CODE REVIEW

File: test/apps/iodemo/io_demo.cc [hunk 1]
```diff
@@ -278,6 +278,25 @@ class P2pDemoCommon : public UcxContext {
         TERMINATE_SIGNALED
     } status_t;
 
+    bool init() {
+        bool ret = UcxContext::init();
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc]: if (!UcxContext::init()) {
 return false;
}

**Comment by yosefe** [comment]
@Artemy-Mellanox can you pls fill the "Why" description for this PR?

**Comment by yosefe** [comment]
maybe to be on the safe side, clear all signal handlers from the cleanup thread?

File: src/uct/ib/base/ib_device.c [hunk 2]
```diff
@@ -420,15 +420,22 @@ static int uct_ib_device_cleanup_proc(void* arg)
 {
     uct_ib_device_nb_close_ctx *ctx = arg;
     static const char *process_name = "ucx_cleanup";
+    void *dfl_action = NULL;
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c]: can use SIG_DFL?

File: src/uct/ib/base/ib_device.c [hunk 3]
```diff
@@ -420,15 +420,22 @@ static int uct_ib_device_cleanup_proc(void* arg)
...
     char dummy;
-    int fd;
+    int i;
+
+    dfl.sa_handler = SIG_DFL;
+
+    for (i = 1; i < 64; i++) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c]: can go upto SIGUSR2?

File: src/uct/ib/base/ib_device.c [hunk 4]
```diff
@@ -420,13 +420,11 @@ static int uct_ib_device_cleanup_proc(void* arg)
 {
     uct_ib_device_nb_close_ctx *ctx = arg;
     static const char *process_name = "ucx_cleanup";
-    void *dfl_action = NULL;
+    void *dfl                       = SIG_DFL;
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c]: shouldn't we pass struct sigaction?

Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_device.c]: Yep, https://elixir.bootlin.com/linux/v4.4/source/arch/x86/include/uapi/asm/signal.h#L102

Comment (yosefe) [reply src/uct/ib/base/ib_device.c]: 1. can we use SYS_signal?
2. explain why we pass only one field 
3. change type from void* to  __sighandler_t

File: src/uct/ib/base/ib_device.c [hunk 5]
```diff
@@ -416,16 +416,24 @@ ucs_status_t uct_ib_device_query(uct_ib_device_t *dev,
     return UCS_OK;
 }
 
+struct uct_ib_sigaction {
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:420]: 1. document this struct - does it correspond to kernel ABI?
2. indent on column

File: src/uct/ib/base/ib_device.c [hunk 6]
```diff
@@ -416,16 +416,24 @@ ucs_status_t uct_ib_device_query(uct_ib_device_t *dev,
...
     static const char *process_name = "ucx_cleanup";
-    void *dfl                       = SIG_DFL;
+    struct uct_ib_sigaction dfl     = { SIG_DFL };
     char dummy;
     int i;
 
     for (i = 1; i <= SIGUSR2; i++) {
-        ucs_syscall_raw4(SYS_rt_sigaction, i, (long)&dfl, 0, sizeof(dfl));
+        ucs_syscall_raw4(SYS_rt_sigaction, i, (long)&dfl, 0,
+                         sizeof(dfl.uapi_sa_mask));
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:438]: weird

Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_device.c:438]: https://elixir.bootlin.com/linux/v4.4/source/kernel/signal.c#L3307

Comment (yosefe) [reply src/uct/ib/base/ib_device.c:438]: ic, it should be the size of sigset_t and not size of whole struct

**Review by yosefe — APPROVED** [review]
pls squash

**Pull request merged**